protocols/consensus-gossip/tuple-space.md

```diff
@@ -0,0 +1,192 @@
+
+In both approaches for synchronization, the tuple space could grow indefinitely, given that the number of heights and rounds is infinite.
+To save memory, entries should be removed from the tuple space as soon as they become stale, that is, they are no longer useful.
+For example, if a new height is started, all entries corresponding toe previous heights become stale.
+
+In general, simply forgetting stale entries in the local view would save the most space.
+However, it could lead to entries being added back or circulating on the network.
+Even if re-adding tuples does not compromise CometBFT's correctness, removed entries should not be added again, for performance and resource utilization sake.
+
+Re-adding entries may be prevented by keeping _tombstones_ for the removed entries.
+We note the tombstone for an entry $e$ as $\bar{e}$.
+
+With time, even small tombstones may accrue and need to be garbage collected, in which cause the corresponding entry may be added again; again, this will not break correctness and as long as tombstones are kept for long enough, the risk of re-adding is minimal.
+
+In the case of the Tendermint algorithm we note that staleness comes from adding newer messages (belonging to higher rounds and heights) to the tuple space.
+Hence, in Approach Two, if as an optimization these newer messages are exchanged first, then the stale messages can be excluded before being shared to other nodes that might have forgotten them and tombstones may not be needed at all.
+
+
+### Querying the Tuple Space
+
+The tuple space is consulted through queries, which have the same form as the entries.
+Queries return all entries whose values match those in the query, where a `*` matches all values.
+For example, suppose a node's local view of the tuple space has the following entries, here organized as rows of a table for easier visualization:
+
+| Height | Round | Step     | Validator | Value |
+|--------|-------|----------|-----------|-------|
+| 1      | 0     | Proposal | p         | pval  |
+| 1      | 0     | PreVote  | p         | vval  |
+| 1      | 1     | PreCommit| q         | cval  |
+| 2      | 0     | Proposal | p         | pval' |
+| 2      | 2     | PreVote  | q         | vval' |
+| 2      | 3     | PreCommit| q         | cval' |
+
+- Query $\lang 0, 0, Proposal, p, * \rang$ returns $\{ \lang 0, 0, Proposal, p, pval \rang \}$
+- Query $\lang 0, 0, *, p, * \rang$ returns $\{ \lang 0, 0, Proposal, p, pval \rang,  \lang 0, 0, PreVote, p, vval \rang \}$.
+
+
+#### Local views
+
+Let $T$ be the tuple space and $t_p$ be node $p$'s approximation of $T$, also denominated $p$'s **local views**.
+Because of the asynchronous nature of distributed systems, local views may not include entries in the space or may still include entries no longer in the space.
+
+Formally, let $P$ be the set of validators; $T = \cup_P t_p$ and $e \in T \Leftrightarrow \exists p \in P, e \in t_p \land \not\exists q \in Q, \bar{e}\in t_q$
```

Comment:
Not sure what you want to say with the formula.

Reply:
T is the actual tuple space, the global view if you will, and combines the values of all local views.
No node can see it and I've defined it here expecting to be useful from a formalization point of view. I may be exactly the case for dealing with byzantine validators. That is, byzantine nodes can poison local views with tuples not with the global view (because the collide). But I will need to think more about it before committing to a solution.
Saving your comments on a todo on the followup PR.

Reply:
$Q$ has not been defined. I assume it should be $P$ instead.

Reply:
I think I commented something regarding that on the companion PR on the Comet repository.

You are assuming that is enough a single process to "tombstone" a tuple to render that tuple not part of the global set of tuples. I have the impression that this is too strong, in particular in light of Byzantine behavior.

Reply:
Indeed. I think we can get away with it by including proofs of staleness on tombstones or making tombstones local only.